src/main/java/org/reactome/addlinks/db/ReferenceObjectCache.java

```diff
@@ -106,183 +106,217 @@ private static synchronized void buildReferenceCaches(String className, Map<Stri
+					(!className.equals(ReactomeJavaConstants.ReferenceMolecule)
+							&& !className.equals(ReactomeJavaConstants.DatabaseIdentifier)
+							&& !className.equals(ReactomeJavaConstants.ReferenceTherapeutic)
+						? cacheBySpecies.size() : "N/A"),
+					objectCacheByIdentifier.size(),
+					cacheByID.size()
+				);
+	}
+
+	/**
+	 * Adds an object to a cache that caches by identifier.
+	 * @param className - type of object being cached.
+	 * @param objectCacheByIdentifier - the cache to add the object into.
+	 * @param referenceObject - The object that will be added to a cache.
+	 * @throws Error
+	 */
+	private static void addObjectToIdentifierCache(String className, Map<String, List<GKInstance>> objectCacheByIdentifier, GKInstance referenceObject) throws Error
+	{
+		try
+		{
+			if (referenceObject.getSchemClass().isValidAttribute(ReactomeJavaConstants.identifier) && referenceObject.getAttributeValue(ReactomeJavaConstants.identifier) != null)
+			{
+				String identifier = (String) referenceObject.getAttributeValue(ReactomeJavaConstants.identifier);
 				try
 				{
-					refDBID = String.valueOf(((GKInstance) referenceObject.getAttributeValue(ReactomeJavaConstants.referenceDatabase)).getDBID());
-					if (! cacheByRefDB.containsKey(refDBID))
+					if (!objectCacheByIdentifier.containsKey(identifier))
 					{
-						List<GKInstance> byRefDb = Collections.synchronizedList( new LinkedList<GKInstance>() );
-						byRefDb.add(referenceObject);
-						cacheByRefDB.put(refDBID, byRefDb);
+						objectCacheByIdentifier.put(identifier, Collections.synchronizedList(new ArrayList<GKInstance>(Arrays.asList(referenceObject))) );
 					}
 					else
 					{
-						cacheByRefDB.get(refDBID).add(referenceObject);
+						objectCacheByIdentifier.get(identifier).add(referenceObject);
 					}
 				}
-				catch (InvalidAttributeException e)
+				catch (ArrayIndexOutOfBoundsException e)
 				{
-					logger.error("Could not get the \"{}\" attribute for {} because it not valid for this object.", ReactomeJavaConstants.referenceDatabase, referenceObject);
-				}
-				catch (Exception e)
-				{
-					logger.error("Could not get the \"{}\" attribute for {}. Reason: {}", ReactomeJavaConstants.identifier, referenceObject, e.getMessage());
+					logger.error("ArrayIndexOutOfBounds was caught! Identifier that triggered this was: {}; className: {}; objectCacheByIdentifier has {} items; objectCacheByIdentifier[identifier] has {} items", identifier, className, objectCacheByIdentifier.size(), objectCacheByIdentifier.containsKey(identifier) ? objectCacheByIdentifier.get(identifier).size() : " *THAT IDENTIFIER IS NOT IN THAT MAP!* ");
 					e.printStackTrace();
+					throw new Error(e);
 				}
+
+			}
+		}
+		catch (InvalidAttributeException e)
+		{
+			logger.error("Object {} does not have an identifier attribute, error: {}", referenceObject, e.getMessage());
+			e.printStackTrace();
+		}
+		catch (Exception e)
+		{
+			e.printStackTrace();
+			throw new Error(e);
```

Comment:
Is there a need to print the stacktrace if the error is rethrown (i.e. once the error is handled by Java, a stacktrace should be printed)?

Reply:
This is old code so I'm guessing that the exception was getting caught and the stack trace was _not_ getting printed, for some reason.